### DISCUSSION TITLE
How to prevent re-render at root level when using isDirty?

### DISCUSSION THREAD

**Discussion by kiocos** [react-hook-form/react-hook-form#10314] [Q&A] [answered]
Following up with [this issue](https://github.com/react-hook-form/react-hook-form/issues/10312), I have an undesired re-render at root level due to change in isDirty. Bill told me " if re-render at root is not desired, use useFormState"

[This example](https://codesandbox.io/s/useformstate-75xly) from the documentation uses useFormState and dirtyFields, however the re-render at root level still occurs. I'm opening this thread because I'm not sure how I should use useFormState to prevent re-render at the root level when isDirty changes.

Ideally, I'd like to re-render just a submit button when isDirty changes and not the every form input.

Thanks a lot!!

**Comment by leapful** [comment, marked as answer]
@kiocosta 

It seems you are mis-understanding how React hook and re-rendering works.

In you codesandbox sample, you are using `useFormState` at the root component level, so every time hook `useFormState` (or any React hook) had an update, it will trigger a re-rendering on the component that is currently using it, a.k.a your root App component and leads to your problem "re-render at root level still occurs".

If you want to avoid this behavior, you move `useFormState` to an isolate component that needs to be re-rendered following the hook instead of root level component. In your case, it's the submit button.

Working example: https://codesandbox.io/s/useformstate-forked-ndm489?file=/src/Submit.tsx
